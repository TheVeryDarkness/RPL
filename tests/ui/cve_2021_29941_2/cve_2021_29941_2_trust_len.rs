//@compile-flags: -Z inline-mir=false

// #[rpl::dump_mir(dump_cfg, dump_ddg)]
pub fn swap_index(bla: impl ExactSizeIterator<Item = u32>) -> Vec<u32> {
    let mut vec = Vec::new();
    let len = bla.len();
    // vec.reserve(len);
    // let arr: &mut [u32] = unsafe { std::slice::from_raw_parts_mut(vec.as_mut_ptr(), bla.len()) };
    // for (i, a) in bla.enumerate() {
    //     arr[a as usize] = i as u32;
    // }

    unsafe {
        vec.set_len(len);
        //~^ERROR: it is unsound to trust return value of `std::iter::ExactSizeIterator::len` and pass it to an unsafe function like `std::vec::Vec::set_len`, which may leak uninitialized memory
    }
    vec
}

fn main() {}
