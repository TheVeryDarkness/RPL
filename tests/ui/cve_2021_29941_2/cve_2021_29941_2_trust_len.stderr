error: it is unsound to trust return value of `std::iter::ExactSizeIterator::len` and pass it to an unsafe function like `std::vec::Vec::set_len`, which may leak uninitialized memory
  --> tests/ui/cve_2021_29941_2/cve_2021_29941_2_trust_len.rs:14:9
   |
LL |     let len = bla.len();
   |               --------- `std::iter::ExactSizeIterator::len` used here
...
LL |         vec.set_len(len);
   |         ^^^^^^^^^^^^^^^^
   |
   = help: incorrect implementation of `std::iter::ExactSizeIterator::len` must not cause safety issues

error: aborting due to 1 previous error

